SecurityPkg: Add measurement of Firmware Debugger Enabled based on DeviceState

VivianNK opened the pull request:
## Description

Add measurement of Firmware Debugger
Enabled based on DeviceState.

Added both to Tcg2Pei and Tcg2Dxe. The measurement is redundant
in Tcg2Dxe, but is added for consistency with previous functionality.
The plan is to remove the PcdFirmwareDebuggerInitialized PCD
and replace its usage with the DeviceStateLib

This will prevent the system from booting if the device is in an
an insecure state, as determined by the DeviceStateLib from
MdeModulePkg.

- [x] Impacts functionality?
- [x] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested
Local CI

## Integration Instructions
N/A

os-d commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.c:
```diff
@@ -42,14 +42,15 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 // MU_CHANGE [END]
-// MU_CHANGE_131467
-// MU_CHANGE [BEGIN] - Move to 256-bit PCRs.
+// MU_CHANGE [BEGIN] - 131467 - Move to 256-bit PCRs.
```
I would not update these two comments above, if we are going to do that, we should do it in the next integration and squash it in with the commit and not pollute this one

os-d commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.c:
```diff
@@ -42,14 +42,15 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #include <Library/Tcg2PreUefiEventLogLib.h>
 // MU_CHANGE [END]
+// MU_CHANGE [BEGIN]- Add InsecureDeviceStateString
+#define MSFT_INSECURE_DEVICE_STATE_STRING  "Microsoft UEFI Insecure Device State"
```
I think we can generalize this away from Microsoft, that original code that did this was old and specific to us, this is theoretically something that could be leveraged by others.

os-d commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.c:
```diff
@@ -644,6 +645,64 @@ MeasureCRTMVersion (
+  //
+  // Use FirmwareVersion string to represent CRTM version.
+  // OEMs should get real CRTM version string and measure it.
+  //
```
This comment does not seem to be in line with what is happening

os-d commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.c:
```diff
@@ -644,6 +645,64 @@ MeasureCRTMVersion (
+           (UINT8 *)String
+           );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "Failed to measure Insecure Device State to PCR 07\n"));
```
Might as well throw an assert in here. We could consider panicking, too. @cfernald @makubacki thoughts?

VivianNK replied:
Should we return from the function or still try to measure into PCR 0?

VivianNK replied:
Talked offline, determined we will panic and not cap the PCR as it's not a PCR-capping error scenario defined in the specification.

os-d commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.c:
```diff
@@ -1068,11 +1127,18 @@ PeimEntryMP (
+  if (GetDeviceState() & GetInsecuteDeviceStateSetting()) {
+    Status = MeasureInsecureDeviceState ();
+    // TODO we don't want to do anything if the measurement fails? Just override the status with the next measurement?
+    // E.g. could fail here but succeed the last measurement and return success. 
```
I would at minimum assert and consider panicking, based on the conversation above.

makubacki commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.c:
```diff
@@ -1068,11 +1127,18 @@ PeimEntryMP (
   // MU_CHANGE [END]
 
+  // MU_CHANGE [BEGIN] - Measure Insecure Device State into PCR07 and PCR04
+  if (GetDeviceState() & GetInsecuteDeviceStateSetting()) {
```
I assume `Insecute` is a typo?

VivianNK replied:
yes, fixed. WIP so still need to test

VivianNK commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.c:
```diff
@@ -1073,6 +1113,23 @@ PeimEntryMP (
+      ASSERT_EFI_ERROR (Status);
+
+      PanicReport (__FILE__, __LINE__, "Failed to measure Firmware Debugger Enabled!\n");
+      // TODO cap PCR?
```
@makubacki Please advise. Would it make sense to cap the PCR here? How is that done?

os-d replied:
My thought is that we would want to cap PCR 7 and not panic? One use case is the system is under debug legitimately and so you are okay with PCR 7 breaking, but still want to continue booting to debug whatever issue (which might be why the measurement failed :). But I would defer to @makubacki

VivianNK replied:
Talked offline - see other comment.

os-d commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.c:
```diff
@@ -644,6 +648,38 @@ MeasureCRTMVersion (
+  TcgEvent.EventSize = sizeof (FIRMWARE_DEBUGGER_EVENT_STRING) - 1;
+
+  DEBUG ((DEBUG_INFO, "Measuring Device State: Firmware Debugger Enabled\n"));
+  return TcgDxeHashLogExtendEvent (
```
This needs to be the PEI version of this function

os-d commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.c:
```diff
@@ -644,6 +648,38 @@ MeasureCRTMVersion (
+{
+  TCG_PCR_EVENT_HDR  TcgEvent;
+
+  TcgEvent.PCRIndex  = 7;
```
Are we still planning on also measuring this to PCR 0?

VivianNK replied:
In a separate commit not to be upstreamed, I think yes. @cfernald, right, we wanted to use PCR 0 to indicate the code is in an unknowable state?

VivianNK replied:
Offline we determined if we want to measure into PCR 0, it should be in a separate PR

Flickdm commented on SecurityPkg/Tcg/Tcg2Dxe/Tcg2Dxe.c:
```diff
@@ -2497,14 +2501,32 @@ MeasureSecureBootPolicy (
+      ASSERT_EFI_ERROR (Status);
+
+      PanicReport (__FILE__, __LINE__, "Failed to measure Firmware Debugger Enabled!\n");
+      // TODO cap PCR?
```
Handle TODO?

VivianNK replied:
I did the thing where I commit amend and push, but didn't actually add the files I wanted to commit :)

os-d commented on SecurityPkg/Tcg/Tcg2Dxe/Tcg2Dxe.c:
```diff
@@ -2497,14 +2501,32 @@ MeasureSecureBootPolicy (
-  VOID        *Protocol;
+  EFI_STATUS    Status;
+  VOID          *Protocol;
+  DEVICE_STATE  CurrentDeviceState; // MU_CHANGE - Measure Firmware Debugger Enabled
```
nit: to avoid reformatting the existing ones and adding to the MU_CHANGE, an extra line can be added here to avoid the reformatting and potential future rebase failures

os-d commented on SecurityPkg/Tcg/Tcg2Dxe/Tcg2Dxe.c:
```diff
@@ -2497,14 +2501,32 @@ MeasureSecureBootPolicy (
+    Status = MeasureLaunchOfFirmwareDebugger ();
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "Failed to measure Firmware Debugger Enabled!\n"));
+      ASSERT_EFI_ERROR (Status);
```
If we are going to panic, asserting doesn't buy us much here :)

VivianNK replied:
My thought was, if we are in debug mode, assert, otherwise if we're in release mode, we want to panic. 

Are you saying we should panic and not assert at all in both scenarios?

os-d replied:
From a debug point of view, assert vs panic don't make a difference that I'm aware of. You'll stop execution with a message about where you failed. The release build is where the difference is. It's possible the breakpoint vs deadloop functionality of asserts isn't present in panic, but I don't think we need to do both.

os-d commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.c:
```diff
@@ -1060,7 +1097,11 @@ PeimEntryMP (
-  EFI_STATUS  Status;
+  // MU_CHANGE [BEGIN] - Measure Firmware Debugger Enabled
+  EFI_STATUS    Status;
+  DEVICE_STATE  CurrentDeviceState;
```
Same nit here

os-d commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.c:
```diff
@@ -1073,6 +1114,23 @@ PeimEntryMP (
+    Status = MeasureFirmwareDebuggerEnabled ();
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "Failed to measure Firmware Debugger Enabled!\n"));
+      ASSERT_EFI_ERROR (Status);
```
Same

os-d commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.c:
```diff
@@ -1073,6 +1114,23 @@ PeimEntryMP (
+      ASSERT_EFI_ERROR (Status);
+
+      PanicReport (__FILE__, __LINE__, "Failed to measure Firmware Debugger Enabled!\n");
+      // TODO cap PCR?
```
Also removing the TODO

VivianNK commented on SecurityPkg/Tcg/Tcg2Pei/Tcg2Pei.inf:
```diff
@@ -62,8 +62,10 @@
 ## MU_CHANGE_103691
 ## MU_CHANGE [BEGIN] - Add the Tcg2PreUefiEventLogLib
   Tcg2PreUefiEventLogLib
-## MU_CHANGE [END]
```
Need to add this line back

makubacki commented on SecurityPkg/Tcg/Tcg2Dxe/Tcg2Dxe.inf:
```diff
@@ -97,7 +101,7 @@
 [Protocols]
   gEfiTcg2ProtocolGuid                               ## PRODUCES
-  gMuTcg2ProtocolExGuid                              ## PRODUCES # MU_CHANGE - Add a new protocol to support Log-only events.
+  gMuTcg2ProtocolExGuid                              ## PRODUCES # MU_CHANGE - Add a new protocol to support Log-only events. 
```
Remove trailing whitespace added.

```suggestion
  gMuTcg2ProtocolExGuid                              ## PRODUCES # MU_CHANGE - Add a new protocol to support Log-only events.
```

VivianNK committed:
SecurityPkg: Add measurement of Firmware Debugger
Enabled based on DeviceState.

Added both to Tcg2Pei and Tcg2Dxe. The measurement is redundant
in Tcg2Dxe, but is added for consistency with Tcg2Pei.
The plan is to remove the PcdFirmwareDebuggerInitialized PCD
and replace its usage with the DeviceStateLib

This will prevent the system from booting if the device is in an
an insecure state, as determined by the DeviceStateLib from
MdeModulePkg.

Signed-off-by: Vivian Nowka-Keane <vnowkakeane@microsoft.com>
